### PR TITLE
[pipeline]: fix p2p comm, add metadata cache and support llama interleaved pp

### PR DESCRIPTION
**Pull request by CWHer**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/pipeline/stage_manager.py [hunk 1]
```diff
@@ -174,3 +169,10 @@ def init_process_group_by_stages(self, stages: List[int]) -> ProcessGroup:
             ProcessGroup: Process group of the given stages.
         """
         return self.pg_mesh.get_group_along_axis(self.pipeline_axis, stages)
+
+    @contextlib.contextmanager
+    def set_model_chunk_id(self, model_chunk_id: int):
```
Comment (ver217) [inline colossalai/pipeline/stage_manager.py:174]: `switch_model_chunk_id()` may be clearer

Comment (CWHer) [reply colossalai/pipeline/stage_manager.py]: 3b84204 fix this.

File: colossalai/pipeline/p2p.py [hunk 2]
```diff
@@ -286,52 +309,79 @@ def _send_recv_serialization_object(
...
+    Args:
+        object (Any): object needed to be sent
+        send_dst (int): rank of the destination
+        recv_src (int): rank of the source
+        send_group (ProcessGroup, optional): process group of sender
+        recv_group (ProcessGroup, optional): process group of receiver
+        send_metadata (bool, optional): whether to send metadata
+        metadata_recv (P2PMetadata, optional): metadata of the object to be received
+    """
+    if send_dst is None and recv_src is None:
```
[the remark this answers is not in the document]
Comment (CWHer) [reply colossalai/pipeline/p2p.py]: 3b84204 fix this.

File: colossalai/pipeline/p2p.py [hunk 3]
```diff
@@ -286,52 +309,79 @@ def _send_recv_serialization_object(
...
+    #   we are not able to do that (1. send & recv metadata 2. send & recv). So we need to split the send & recv into two parts in this case.
+    if (send_dst is not None and recv_src is not None) and (send_metadata or metadata_recv is None):
+        warnings.warn("Fall back to individual send & recv")
+        _communicate(object, send_dst=send_dst, recv_src=None, send_group=send_group, send_metadata=send_metadata)
+        return _communicate(None, send_dst=None, recv_src=recv_src, recv_group=recv_group, metadata_recv=metadata_recv)
+
+    # NOTE: only the following 5 cases are valid:
+    #   1. send() [needs extra metadata] and no recv()
+    #   2. recv() [needs extra metadata] and no send()
+    #   3. neither send() or recv() need extra metadata
```
[the remark this answers is not in the document]
Comment (CWHer) [reply colossalai/pipeline/p2p.py]: 3b84204 fix this.

File: colossalai/pipeline/p2p.py [hunk 4]
```diff
@@ -286,52 +309,79 @@ def _send_recv_serialization_object(
...
+    assert recv_src is None or recv_group is not None, "recv_group must be specified when recv_src is not None"
+    send_metadata = send_metadata or (object is not None and not _check_if_fast_send_available(object))
+    assert (
+        metadata_recv is None or metadata_recv.data_type != P2PDataType.Serialization
+    ), "metadata_recv type must not be Serialization"
+
+    # NOTE: send & recv should be atomic operations. However, if we need to send metadata or receive metadata,
+    #   we are not able to do that (1. send & recv metadata 2. send & recv). So we need to split the send & recv into two parts in this case.
+    if (send_dst is not None and recv_src is not None) and (send_metadata or metadata_recv is None):
+        warnings.warn("Fall back to individual send & recv")
```
[the remark this answers is not in the document]
Comment (CWHer) [reply colossalai/pipeline/p2p.py]: 3b84204 fix this.

**Comment by CWHer** [comment]
`torchrun --master_port=1911 --nproc_per_node=4 benchmark.py \
    -c 7b -p 3d --grad_checkpoint --max_length 1024 \
    --tp 2 --pp 2 --zero 0 -b 384 --mbs 24 -s 3
image

File: colossalai/pipeline/schedule/interleaved_pp.py [hunk 5]
```diff
@@ -27,15 +28,25 @@ def __init__(
         assert (
             num_microbatch is not None or microbatch_size is not None
         ), "Either num_microbatch or microbatch_size should be provided"
+        self.logger = get_dist_logger()
+        self.logger.warning("If pipeline hangs, please enlarge NCCL_BUFFSIZE.")
```
Comment (ver217) [inline colossalai/pipeline/schedule/interleaved_pp.py:32]: Can we enlarge `NCCL_BUFFSIZE` automatically?

File: colossalai/pipeline/schedule/one_f_one_b.py [hunk 6]
```diff
@@ -42,14 +43,24 @@ def __init__(
         assert (
             num_microbatches is not None or microbatch_size is not None
         ), "Either num_microbatches or microbatch_size should be provided"
+        self.logger = get_dist_logger()
+        self.logger.warning("If pipeline hangs, please enlarge NCCL_BUFFSIZE.")
```
Comment (ver217) [inline colossalai/pipeline/schedule/one_f_one_b.py:47]: same

File: colossalai/pipeline/schedule/interleaved_pp.py (hunk 5 above)
Comment (CWHer) [reply colossalai/pipeline/schedule/interleaved_pp.py]: 15fdb26 fix this.

**Review by ver217 — APPROVED** [review]

**Pull request merged**